Fixed vr inputs showing up as gamepad inputs

sinnstudio-nas opened the pull request:
Fixed an issue where vr inputs were showing up as gamepad inputs.

Before:

![image](https://user-images.githubusercontent.com/50789023/110894106-e1483380-82bc-11eb-9b06-5399018830e7.png)

After:

![image](https://user-images.githubusercontent.com/50789023/110894134-ee652280-82bc-11eb-992f-63c7ea70a6f8.png)

goncasmage1 commented:
Looks good, however, I think you'll have to expand that category to not just gamepad but also all the platform-specific gamepads, like PS4, XBox, Switch, etc. (or just check that it isn't any of the VR platform categories)

sinnstudio-nas commented:
Hey @goncasmage1 ,

Updated the PR. Added a helper to check if it is VRkey or not.

2 issues remain:
- This works for console BUT... if you are on PS4... you could see xbox or other platform keys.
- Need to figure out a way to display only VR keys and by platform. Let's say... I am on Oculus... I should only see Oculus keys.

Both of these issues go back to the same problem... we need to be able to control what keys to display at run-time somehow

goncasmage1 commented:
Hey,

Yeah those issues are something I'll have to look into in a following task in the backlog.
However, there is one more change I'd like you to make, which is to add the VR entry to the InputRestrictions array. Right now, the RespectsRestriction function will not allow any VR inputs with the changes you made. So I'd appreciate it if you also did that change.

Thanks 💪

sinnstudio-nas commented on Source/UINavigation/Private/UINavBlueprintFunctionLibrary.cpp:
```diff
@@ -95,8 +99,22 @@ bool UUINavBlueprintFunctionLibrary::RespectsRestriction(FKey Key, EInputRestric
+				return (Key.IsGamepadKey() && !IsVRKey(Key));
+			}
+			else {
+				return ((Key.IsGamepadKey() || Key.GetMenuCategory().ToString() == Platform) && !IsVRKey(Key));
```
Wrong I need to check if it's a gamepad key or a console category key... as consoles have a mixture

goncasmage1 replied:
Won't this IsGamepadKey return true even if the key is specifically of the PS4 controller or XBox controller for instance?

sinnstudio-nas replied:
You know what... it probably will. I will test out shortly.

sinnstudio-nas replied:
Fixed. You were right.

sinnstudio-nas commented on Source/UINavigation/Private/UINavBlueprintFunctionLibrary.cpp:
```diff
@@ -95,8 +99,22 @@ bool UUINavBlueprintFunctionLibrary::RespectsRestriction(FKey Key, EInputRestric
+			}
+			else if (HMD == "PSVR") {
+				return IsKeyInCategory(Key, "PSMove");
+			}
```
No, As for SteamVR, it should stay unbound

goncasmage1 replied:
How about the Vive and Windows Mixed Reality?

sinnstudio-nas replied:
All run under SteamVR only.

sinnstudio-nas commented on Source/UINavigation/Private/UINavBlueprintFunctionLibrary.cpp:
```diff
@@ -81,6 +83,8 @@ void UUINavBlueprintFunctionLibrary::ResetInputSettings()
 
 bool UUINavBlueprintFunctionLibrary::RespectsRestriction(FKey Key, EInputRestriction Restriction)
 {
+	FString HMD = GEngine->XRSystem->GetSystemName().ToString();
```
Hey @goncasmage1 ,

Beginplay cannot be called in a functionlibrary since it's not an actor. We can add a getter function but for now I declared these variables at the top of the file

goncasmage1 commented on Source/UINavigation/Private/UINavBlueprintFunctionLibrary.cpp:
```diff
@@ -6,6 +6,11 @@
+#include "IXRTrackingSystem.h"
+
+FString Platform = UGameplayStatics::GetPlatformName();
+FString HMD = GEngine->XRSystem->GetSystemName().ToString();
```
Add static to both these variables

goncasmage1 commented on Source/UINavigation/Private/UINavBlueprintFunctionLibrary.cpp:
```diff
@@ -6,6 +6,11 @@
+#include "IXRTrackingSystem.h"
+
+FString Platform = UGameplayStatics::GetPlatformName();
+FString HMD = GEngine->XRSystem->GetSystemName().ToString();
```
I don't think this is a very safe thing to do actually. Sorry for flipping back on this but this should probably go back to the RespectsRestriction function

sinnstudio-nas replied:
Fixed
